### PR TITLE
Switch to perl memory allocation for better compatibility

### PR DESCRIPTION
**Pull request by toddr**
We have a use case coming where it would be helpful to have Class::XSAccessor use perl memory management as opposed to glibc memory management. Would you be open to this change?

Thanks,
Todd

### CODE REVIEW

**Comment by toddr** [comment]
See this file for the net change we're proposing here

https://github.com/tsee/Class-XSAccessor/pull/6/files#diff-b5f092c9429ad42f0835ac748cfc26d4

**Comment by atoomic** [comment]
I can see several advantages there:
1. get rid of the c functions by directly using the macros
2. increase portability on other systems by using Perl memory functions
3. allow static setup when compiled

**Comment by toddr** [comment]
I've re-pushed this commit. It's open for comments/questions, etc. I tested on my own travis and the change seems to work across the board.

**Comment by toddr** [comment]
In effect you can see the change being made by looking at the changes in this header file: https://github.com/tsee/Class-XSAccessor/pull/6/files#diff-b5f092c9429ad42f0835ac748cfc26d4

File: XSAccessor.xs [hunk 1]
```diff
@@ -369,8 +369,8 @@ STMT_START {                                                                 \
 STMT_START {                                                                 \
   autoxs_hashkey *hk_ptr = get_hashkey(aTHX_ obj_hash_key, obj_hash_key_len);\
   INSTALL_NEW_CV_WITH_PTR(name, xsub, hk_ptr);                               \
-  hk_ptr->key = (char*)cxa_malloc(obj_hash_key_len+1);                       \
-  cxa_memcpy(hk_ptr->key, obj_hash_key, obj_hash_key_len);                   \
+  cxa_malloc(hk_ptr->key, obj_hash_key_len+1, char);                         \
```
Comment (chocolateboy) [inline XSAccessor.xs]: Trailing space.

File: cxsa_hash_table.c [hunk 2]
```diff
@@ -126,8 +126,8 @@ CXSA_HashTable_grow(HashTable* table) {
     UV newsize = oldsize * 2;
     UV i;
 
-    array = (HashTableEntry**)cxa_realloc((void*)array, sizeof(HashTableEntry*)*newsize);
-    cxa_memzero(&array[oldsize], (newsize-oldsize)*sizeof(HashTableEntry*));
+    cxa_realloc(array, newsize, HashTableEntry*);
+    cxa_memzero(&array[oldsize], newsize-oldsize, HashTableEntry*);
```
Comment (chocolateboy) [inline cxsa_hash_table.c]: `newsize-oldsize` -> `newsize - oldsize`.

File: cxsa_main.c [hunk 3]
```diff
@@ -32,7 +32,7 @@ STATIC
 autoxs_hashkey *
 _new_hashkey() {
   autoxs_hashkey * retval;
-  retval = (autoxs_hashkey *) cxa_malloc( sizeof(autoxs_hashkey) );
+  cxa_malloc( retval, 1, autoxs_hashkey );
```
Comment (chocolateboy) [inline cxsa_main.c]: Extraneous spaces.

File: XSAccessor.xs (hunk 1 above)
Comment (chocolateboy) [reply XSAccessor.xs]: `obj_hash_key_len+1` -> `obj_hash_key_len + 1`. (Yes, some of these are inconsistent in the original.)

File: cxsa_hash_table.c (hunk 2 above)
Comment (chocolateboy) [reply cxsa_hash_table.c]: Spaces around the `*`.

**Comment by toddr** [comment]
I've pushed 3 changes as requested.

**Comment by chocolateboy** [comment]
Thanks. There's no need to force-push all fixes/updates, btw. They can be squashed on merge.

File: cxsa_hash_table.c [hunk 4]
```diff
@@ -99,10 +99,10 @@ CXSA_HashTable_store(HashTable* table, const char* key, STRLEN len, void * value
         entry->value = value;
     } else {
         const UV index = CXSA_string_hash(key, len) & (table->size - 1);
-        entry = (HashTableEntry*)cxa_malloc(sizeof(HashTableEntry));
+        cxa_malloc(entry, 1, HashTableEntry);
+        cxa_malloc( entry->key, (len+1), char );
 
-        entry->key = (char*)cxa_malloc( (len+1) );
-        cxa_memcpy((void*)entry->key, (void*)key, len+1);
+        cxa_memcpy((char*) entry->key, (char*) key, len+1, char);
```
Comment (chocolateboy) [inline cxsa_hash_table.c]: Two more: `len+1` -> `len + 1`.

**Comment by toddr** [comment]
In all honesty, I'd actually recommend a patch behind this one to just get rid of cxsa_memory.h It looks like an unnecessary wrapper at this point.

**Comment by toddr** [comment]
Can these be merged?

**Comment by atoomic** [comment]
.. and released?

**Comment by chocolateboy** [comment]
This project is unmaintained. @tsee hasn't replied and I haven't touched it for 7 years. Your best bet is to fork it, merge your changes, and release them as a new module.

Alternatively, if you want to take over maintainership, you'll need to ask Steffen. I don't have admin access here.

**Comment by tsee** [comment]
Happy to pass it off to good hands! LMK!

**Comment by atoomic** [comment]
@tsee or @chocolateboy would it be possible to add co-maint to the CPAN users toddr and atoomic?
Would you prefer keeping that GitHub repo and gives both of us admin permissions (same GitHub usernames) or you prefer that we fork the repo and advertise the new one?

**Comment by tsee** [comment]
I'll do my best to remember to do this tomorrow. Don't be shy to call me
out of I forget. I'm off work this week and nominally not touching the
computer. :)

I'm fairly ambivalent regarding the repository. Happy to add you in any
case and you'll have the choice then!

**Comment by tsee** [comment]
Alright, as promised, I added atoomic+toddr to this repository (github sent you an invitation) and also added you both as co-maintainers on PAUSE. Chocolateboy didn't have co-maint on one of the namespaces, so I fixed that, too. If one of you would prefer to get the primary bits, LMK!

Cheers and sorry for being so unresponsive. I do read my email, it's just that I struggle to keep up and the perl stuff tends not to float to the top these days. I think you all know my various email addresses, so if I'm required in any capacity and don't seem to be responding, never be shy to send me email directly.

**Pull request merged**